examples/amm/src/lib.rs

```diff
@@ -0,0 +1,119 @@
+use async_graphql::{Request, Response};
+use linera_sdk::{
+    base::{ContractAbi, ServiceAbi},
+    views::MapView,
+};
+use linera_views::views::ViewError;
+use serde::{Deserialize, Serialize};
+use thiserror::Error;
+
+pub struct AMMAbi;
+
+impl ContractAbi for AMMAbi {
+    type InitializationArgument = ();
+    type Parameters = ();
+    type Operation = Operation;
+    type ApplicationCall = ();
+    type Message = ();
+    type SessionCall = ();
+    type Response = ();
+    type SessionState = ();
+}
+
+impl ServiceAbi for AMMAbi {
+    type Query = Request;
+    type QueryResponse = Response;
+    type Parameters = ();
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub enum Operation {
+    Swap {
+        input_token: String,
+        output_token: String,
+        amount: u64,
+    },
+    AddLiquidity {
+        token: String,
+        amount: u64,
+    },
+    RemoveLiquidity {
+        token: String,
+        amount: u64,
+    },
+}
+
+pub fn calculate_output_amount(
+    input_amount: u64,
+    input_pool_balance: u64,
+    output_pool_balance: u64,
+) -> Result<u64, AMMError> {
+    if input_pool_balance == 0 || output_pool_balance == 0 {
+        return Err(AMMError::InvalidPoolBalanceError);
+    }
+
+    let input_pool_balance_float = input_pool_balance as f64;
+    let output_pool_balance_float = output_pool_balance as f64;
+
+    let input_amount_float = input_amount as f64;
+    let output_amount_float =
+        (input_amount_float * output_pool_balance_float) / input_pool_balance_float;
+
+    Ok(output_amount_float as u64)
+}
```

Comment:
So the rule is something like: "If you give me 5% of my Foocoin balance, you get 5% of my Barcoin balance"; if the percentages are always small, that would approximate a Constant Product MM? But for large percentages we'd make a loss: E.g. if the input amount is 100% of the input pool balance, I'll just give you _all_ my Barcoin, and the product will end up being 0?

It would seem safer to explicitly keep the product constant and make sure that (ignoring types here):

`(in_pool + in_amt) * (out_pool - out_amt) = in_pool * out_pool`

So:

`out_amt = out_pool - in_pool * out_pool / (in_pool + in_amt)`

Reply:
Let me try to fix it